rfc/rfc-66/rfc-66.md

```diff
@@ -0,0 +1,119 @@
+
+![multi-writer](multi_writer.png)
+
+### The Compaction Procedure
+
+The compaction service is the duty role that actually resoves the conflicts. Within a file group, it sorts the files then merge all the record payloads for a record key.
+The event time sequence is respected by combining the payloads with even time field provided by the payload (known as the `preCombine` field in Hudi).
+
+![compaction procedure](compaction.png)
+
+#### Non-Serial Compaction Plan Schedule
+Currently, the compaction plan scheduling must be in serial order with the writers, that means, while scheduling the compaction plan, no ongoing writers should be writing to
+the table. This restriction makes the compaction almost impossible for multi streaming writers because there is always an instant writing to the table for streaming ingestion.
+
+In order to unblock the compaction plan scheduling and keep the completeness of the readers, we introduce the completion time for file slice generation:
+
+- Support quick look-up from instant time ➝ completion time, the [HUDI-6539](https://issues.apache.org/jira/browse/HUDI-6539) supports fast completion time queries on archived timeline, based on this, we are able
+to support flexible completion time queries on both active and archived timeline, see [HUDI-6725](https://issues.apache.org/jira/browse/HUDI-6725);
+- New compaction plan scheduling to comply with the completion time, that is: only log files that have smaller completion time(than the compaction start instant time) should be considered
+- New file slice generation strategy, a log file with smaller instant time (than the compaction instant time) but greater completion time should be assigned to a new file slice
+- By combining #2 and #3, in general, we are slicing by comparing the compaction start time with the log files completion time.
+
+<img src="non_serial_compaction.png" alt="drawing" width="400"/>
+
+Assumes we have two log files, with instant time & completion time as [t1, t3] and [t2, t5], at t4, a compaction plan was scheduled,
+the plan does not include file t2_t5_v1.w2, in the reader view, the log file should be assigned to a different file slice than the t4 instant time.
+
+#### Global Monotonically Increasing Timestamp
+
+In order to make the time deterministic among cloud storages, we use a logical time generated by a special **TimeGenerator**, see [HUDI-1623](https://issues.apache.org/jira/browse/HUDI-1623) for details.
+
+### The Log File Naming Convention
+
+We use the current instant time instead of the base commit time in the file name,
+so that tasks from different writers can not conflict in file name. We can also parse the file name to fetch the instant time quickly. 
+Finally, the log name is with the following pattern:
+
+```shell
+${uuid}_${instant}.log.${version}_${task_token}
```

Comment:
Can the log file name contain the delta committime as well?

Reply:
Otherwise hou do we map from completiontime toinstant time '

Reply:
> delta committime

you mean the instant time (start time) ?, we will do that.

Reply:
In the picture of [Non-Serial Compaction Plan Schedule], there are two log files named `t2_t5_v2.w2` and `t1_t3_v1.w1`, the log files names contain instant time and completion time. 
But the log name rule here only contain instant time, does not contain completion time.
I suppose the rules here should be followed, and the log file names in the picture are for easier understanding, I guess?

Reply:
Yes i think log file naming needs to be udpated. It can contain both.
`[File Id]_[Base Transaction timestamp]_[Deltacommit instant time].[Log File Extension].[Log File Version]_[File Write Token]`

Reply:
Updated, but I'm -1 for introduing the `base Transaction timestamp` in the file name, no one knows what is that.